gps2val/gpsd_feeder.py

```diff
@@ -20,115 +20,276 @@
+    root_logger.addHandler(console_logger)
+
+
+class ColorFormatter(logging.Formatter):
+    FORMAT = "{time} {{loglevel}} {logger} {msg}".format(
+        time="\x1b[2m%(asctime)s\x1b[0m",  # grey
+        logger="\x1b[2m%(name)s:\x1b[0m",  # grey
+        msg="%(message)s",
+    )
+    FORMATS = {
+        logging.DEBUG: FORMAT.format(loglevel="\x1b[34mDEBUG\x1b[0m"),  # blue
+        logging.INFO: FORMAT.format(loglevel="\x1b[32mINFO\x1b[0m"),  # green
+        logging.WARNING: FORMAT.format(loglevel="\x1b[33mWARNING\x1b[0m"),  # yellow
+        logging.ERROR: FORMAT.format(loglevel="\x1b[31mERROR\x1b[0m"),  # red
+        logging.CRITICAL: FORMAT.format(loglevel="\x1b[31mCRITICAL\x1b[0m"),  # red
+    }
+
+    def format(self, record):
+        log_fmt = self.FORMATS.get(record.levelno)
+        formatter = logging.Formatter(log_fmt)
+        return formatter.format(record)
+
+
+def parse_env_log(env_log, default=logging.INFO):
+    def parse_level(level, default=default):
+        if type(level) is str:
+            if level.lower() in [
+                "debug",
+                "info",
+                "warn",
+                "warning",
+                "error",
+                "critical",
+            ]:
+                return level.upper()
+            else:
+                raise Exception(f"could not parse '{level}' as a log level")
+        return default
+
+    loglevels = {}
+
+    if env_log is not None:
+        log_specs = env_log.split(",")
+        for log_spec in log_specs:
+            spec_parts = log_spec.split("=")
+            if len(spec_parts) == 1:
+                # This is a root level spec
+                if "root" in loglevels:
+                    raise Exception("multiple root loglevels specified")
+                else:
+                    loglevels["root"] = parse_level(spec_parts[0])
+            if len(spec_parts) == 2:
+                logger = spec_parts[0]
+                level = spec_parts[1]
+                loglevels[logger] = parse_level(level)
+
+    if "root" not in loglevels:
+        loglevels["root"] = default
+
+    return loglevels
+
 
 class Kuksa_Client():
 
     # Constructor
     def __init__(self, config):
-        print("Init kuksa client...")
+        log.info("Init kuksa client...")
         if "kuksa_val" not in config:
-            print("kuksa_val section missing from configuration, exiting")
+            log.error("kuksa_val section missing from configuration, exiting")
             sys.exit(-1)
-        provider_config=config['kuksa_val']
+        provider_config = config['kuksa_val']
         self.client = KuksaClientThread(provider_config)
         self.client.start()
-        if str(provider_config.get('protocol')) == 'ws':  # FIXME: and provider_config.get('insecure') != 'True'
-            print("authorizing...")
-            self.client.authorize(str(provider_config.get('token')))
+        token_string = str(provider_config.get('token'))
```

Comment:
I've changed this to tokenfile in new fix for kuksa-client. So not working with this. We should align on this.

Reply:
Yes. The question is - what should be our preferred method for feeders/providers. Shall they also support both file and token, or only token (as specified today in most documentation for feeders/providers). We have some kuksa-client methods that only supports token, not file (like the async API), and if so the feders/providers must for those cases read the token from file. That is ok. But are there any methods that only accept file path?

Reply:
So with the current change in my PR all functions accept tokens but like you mentioned there are some function which not accept tokenfiles like the sync async lib functions